src/transform/src/join_implementation.rs

```diff
@@ -139,6 +134,23 @@ impl JoinImplementation {
                 }
                 available_arrangements[index].sort();
                 available_arrangements[index].dedup();
+                let reverse_project = project
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, c)| (c, i))
+                    .collect::<HashMap<_, _>>();
+                // Eliminate arrangements refering to columns that have been
+                // projected away by surrounding MFPs.
+                available_arrangements[index].retain(|key| {
+                    key.iter()
+                        .all(|k| k.support().iter().all(|c| reverse_project.contains_key(c)))
+                });
+                // Permute arrangements so columns reference what is after the MFP.
+                for key_set in available_arrangements[index].iter_mut() {
+                    for key in key_set.iter_mut() {
+                        key.permute_map(&reverse_project);
+                    }
+                }
```

Comment:
Currently on main, JoinImplementation is designed to lift Filters blocking use of an arrangement. Projects would never block use of an arrangement because ProjectionLifting lifts all projects to the top of the `MirRelationExpr`.

Now that ProjectionPushdown exists, projects may block use of an arrangement, so JoinImplementation should lift Projects as well as Filters.

Instead of writing specialized code to lift just Filters + Projects, I decided to reuse the MapFilterProject code to have JoinImplementation lift Maps as well as Filters+Projects.